pkg/envoy/xds/cell.go

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package xds
+
+import (
+	"github.com/cilium/hive/cell"
+
+	"github.com/cilium/cilium/pkg/endpointstate"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/promise"
+)
+
+// Cell initializes and manages the Envoy xds server.
+var Cell = cell.Module(
+	"envoy-proxy-xds",
+	"Envoy xDS cache and server",
+
+	cell.Provide(newXDS),
+)
+
+type xdsServerParams struct {
+	cell.In
+
+	Lifecycle       cell.Lifecycle
+	RestorerPromise promise.Promise[endpointstate.Restorer]
+}
+
+type XDS interface {
+	NewServer(resourceTypes map[string]*ResourceTypeConfiguration) *Server
+
+	// Restore returns 'true' if a restorer promise exists
+	Restore() bool
+}
+
+type xds struct {
+	// restorerPromise is initialized only if xDS server should wait sending any xDS resources
+	// until all endpoints have been restored.
+	restorerPromise promise.Promise[endpointstate.Restorer]
+}
+
+func (x *xds) Restore() bool {
+	return x.restorerPromise != nil
+}
+
+func newXDS(params xdsServerParams) (XDS, error) {
```

Comment:
why introducing another type `XDS` and not add the "restore functionality" to the existing type `Server` and expose this one instead? It seems as `Server` depends on the new type (it's implementation `xds`) anyway.

This way, the `NewServer` function would become obsolete - as the dependent envoy.xdsserver would already receive an instance. The only thing left would be to set the config on the Server instance where we currently call `NewServer(resourceTypes map[string]*ResourceTypeConfiguration) *Server`.

It's just that I think we already have enough abstractions for the xDS server :D

Reply:
I put my thoughts into a feature branch building on top of yours: https://github.com/cilium/cilium/compare/main...mhofstetter:pr/mhofstetter/xds-await-ep-restoration?expand=1

the two additional commits should help illustrate my thoughts :sweat_smile:

Reply:
I simplified the last commit based on your comments. Also removed adding a cell to xds, as it seemed overkill in retrospect.

Reply:
nice - thanks!